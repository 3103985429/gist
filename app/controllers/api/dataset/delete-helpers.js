const winston = require('winston');
const async = require('async');
const Batch = require('batch');
const cluster = require('cluster');
const { isEmpty } = require('lodash');

const datasource_description = require('../../../models/descriptions');
const Cron = require('../../../models/crons');
const cronHelper = require('../../../boot/cron-init');
const datasource_file_service = require('../../../libs/utils/aws-datasource-files-hosting');

/**
 * Delete the sources.
 *
 * @param {integer|ObjectId} datasetId
 */
const deleteSource = datasetId =>
    new Promise((resolve, reject) => {
        datasource_description
            .findById(datasetId)
            .populate('_team')
            .exec((err, description) => {
                if (err) {
                    return reject(err);
                }
                if (!description) {
                    return reject(`Datasource id ${datasetId} not found`);
                }

                if (description.apiEndPoint) {
                    description.apiEndPoint = undefined;
                    description.format = undefined;
                    description.JSONPath = undefined;
                    description.save(err => (err ? reject(err) : resolve()));
                } else if (description.connection && description.connection.url) {
                    description.connection = undefined;
                    description.markModified('connection');
                    description.save(err => (err ? reject(err) : resolve()));
                } else if (description.fileName) {
                    const key = `${description._team.subdomain}/datasets/${description.uid}/datasources/${description.uid}_v${description.importRevision}`;

                    datasource_file_service.deleteObject(key, (err, result) => {
                        if (err) {
                            return reject(err);
                        }
                        description.fileName = null;
                        description.save(err => (err ? reject(err) : resolve(result)));
                    });
                }
            });
    });
module.exports.deleteSource = deleteSource;

/**
 * Gets list of dataset to remove
 *
 * @param {ObjectId} datasetId
 */
const findDataset = (datasetId, next) => {
    let descriptions = [];
    datasource_description
        .findById(datasetId)
        .populate('_team')
        .exec((err, data) => {
            if (err) {
                return next(err);
            }

            if (!data) {
                return next(new Error(`No datasource with id: ${datasetId} exists`));
            }

            descriptions.push(data);

            if (!isEmpty(data.previous_datasets)) {
                data.previous_datasets.map(async id => {
                    await datasource_description.findById(id, (err, dataset) => {
                        if (dataset) {
                            descriptions.push(dataset);
                        }
                    });
                });

                next(null, descriptions);
            } else {
                next(null, descriptions);
            }
        });
};
module.exports.findDataset = findDataset;

/**
 * Remove datasource with given schema id.
 *
 * @param {array} descriptions
 * @param {function} next
 */
const removeDatasourceWithSchemaId = (descriptions, next) => {
    async.each(
        descriptions,
        (description, done) => {
            datasource_description
                .find({ $or: [{ schema_id: description._id }, { master_id: description._id }] })
                .populate('_team')
                .exec((err, results) => {
                    if (err) {
                        return next(err);
                    }

                    results.forEach(element => {
                        element.remove();
                    });
                    winston.info(
                        'Removed all the schema descriptions and working draft inherited to the datasource description : ' +
                            description._id
                    );
                    done();
                });
        },
        err => next(err, descriptions)
    );
};
module.exports.removeDatasourceWithSchemaId = removeDatasourceWithSchemaId;

/**
 * Remove other sources.
 *
 * @param {array} descriptions
 * @param {function} done
 */
const removeOtherSources = (descriptions, done) => {
    async.each(
        descriptions,
        (description, next) => {
            datasource_description.find(
                { _otherSources: description._id, 'relationshipFields.by.joinDataset': description._id.toString() },
                (err, docs) => {
                    if (err) {
                        next(err);
                    } else {
                        if (docs.length === 0) {
                            return next();
                        }

                        async.each(
                            docs,
                            (doc, next2) => {
                                let index = doc._otherSources.indexOf(description.uid);
                                doc._otherSources.splice(index, 1);
                                doc.relationshipFields = doc.relationshipFields.filter(field => {
                                    return field.by.joinDataset !== description._id.toString();
                                });
                                doc.dirty = 1;
                                doc.save(next2);
                            },
                            err => {
                                winston.info(
                                    `Removed all the merged description settings inherited to the datasource description: ${description._id}`
                                );
                                next(err);
                            }
                        );
                    }
                }
            );
        },
        err => done(err, descriptions)
    );
};
module.exports.removeOtherSources = removeOtherSources;

/**
 * Remove cron jobs.
 *
 * @param {array} descriptions
 * @param {function} next
 */
const removeCronJobs = (descriptions, next) => {
    async.each(
        descriptions,
        (description, done) => {
            Cron.findOne({ datasourceDescription: description._id }, (err, schedule) => {
                if (err) {
                    return done(err);
                } else if (schedule) {
                    if (cluster.isMaster) {
                        cronHelper.execute('pause', schedule._id, (err, job) => {
                            if (err) {
                                return done(err);
                            } else {
                                winston.info('Stopped the schedule job');
                                schedule.remove(done);
                            }
                        });
                    } else {
                        winston.info('Send to other process to stop the job');
                        process.send({ command: 'pause', id: schedule._id });
                        schedule.remove(done);
                    }
                } else {
                    return done();
                }
            });
        },
        err => next(err, descriptions)
    );
};
module.exports.removeCronJobs = removeCronJobs;
