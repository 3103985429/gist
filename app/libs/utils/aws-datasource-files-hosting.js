const winston = require('winston');
const aws = require('aws-sdk');
const fs = require('fs');
const _ = require('lodash');

const bucket = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const s3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
});
const debugS3 = false;

function _uploadAncillaryFile(filePath, fileName, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getAncillaryAssetKey(teamSubdomin, datasetId, fileName);
    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the ancillary file to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadAncillaryFile = _uploadAncillaryFile;

function _copyAncillaryFile(team_subdomain, originId, destinationId, fileName, callback) {
    const key = getAncillaryAssetKey(team_subdomain, destinationId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getAncillaryAssetKey(team_subdomain, originId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the ancillary file from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyAncillaryFile = _copyAncillaryFile;

function _uploadDataSource(filePath, newFilename, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getDatasourceKey(teamSubdomin, datasetId, newFilename);

    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the datasource to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadDataSource = _uploadDataSource;

function _copySampleDatasource(datasetToDuplicateId, fileName, datasetId, teamSubdomain, callback) {
    const key = getDatasourceKey(teamSubdomain, datasetId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getDatasourceKey('sampleteam', datasetToDuplicateId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the sample datasource from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying sample datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copySampleDatasource = _copySampleDatasource;

function _copyDatasource(team_subdomain, fileName, copyFromId, copyToId, callback) {
    const keyFrom = getDatasourceKey(team_subdomain, copyFromId, fileName, bucket);
    const keyTo = getDatasourceKey(team_subdomain, copyToId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: keyFrom,
        Key: keyTo,
    };

    debug(`Copying the datasource from S3: from ${keyFrom} to ${keyTo}`);

    s3.copyObject(params, function (err, data) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyDatasource = _copyDatasource;

function _getDatasource(description) {
    const key = getDatasourceKey(description._team.subdomain, description.schemaId || description._id, description.fileName);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading the datasource from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getDatasource = _getDatasource;

function _getAncillaryFile(description) {
    const key = getAncillaryAssetKey(description._team.subdomain, description._id, description.ancillaryFile);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading ancillary file from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getAncillaryFile = _getAncillaryFile;

function keyExists(description, callback = _.noop) {
    return new Promise((resolve, reject) => {
        const key = getDatasourceKey(description._team.subdomain, description.schema_id || description._id, description.fileName);

        const params = {
            Bucket: bucket,
            Key: key,
        };
        debug(`Verifying the datasource from S3: ${key}`);

        s3.headObject(params, (err) => {
            if (err) {
                reject(err);
                return callback(err, false);
            }

            resolve();
            callback(null, true);
        });
    });
}

module.exports.keyExists = keyExists;

function _deleteDataset(description, callback) {
    const team_subdomain = description._team.subdomain;
    const key = getDatasetKey(team_subdomain, description.id);
    const param = {
        Bucket: bucket,
        Prefix: key,
    };

    debug(`Removing the datasource from S3: ${key}`);

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Deleted) {
                return callback();
            }

            if (data.Deleted.length === 1000) {
                return _deleteDataset(description, callback);
            }

            callback();
        });
    });
}

module.exports.deleteDataset = _deleteDataset;

function _deleteTeam(subdomain, callback) {
    const param = {
        Bucket: bucket,
        Prefix: getTeamKey(subdomain),
    };

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Contents) {
                return callback();
            }

            if (data.Contents.length === 1000) {
                return _deleteTeam(subdomain, callback);
            }

            callback();
        });
    });
}

module.exports.deleteTeam = _deleteTeam;

function _deleteObject(key, callback) {
    const params = {
        Bucket: bucket,
        Key: key,
    };

    s3.deleteObject(params, function (err, data) {
        if (err) {
            return callback(err);
        }

        callback(null, data);
    });
}

module.exports.deleteObject = _deleteObject;

function getDatasourceKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `datasources/${fileName}`, bucket);
}

function getAncillaryAssetKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `assets/ancillary/${fileName}`, bucket);
}

function getDatasetKey(subdomain, datasetId, resourceName = '', bucket = null) {
    return getTeamKey(subdomain, `datasets/${datasetId}/${resourceName}`, bucket);
}

function getTeamKey(subdomain, resourceName = '', bucket = null) {
    const prefix = bucket ? `${bucket}/` : '';

    return `${prefix}${subdomain}/${resourceName}`;
}

function debug(msg) {
    if (debugS3) {
        winston.debug(msg);
    }
}
