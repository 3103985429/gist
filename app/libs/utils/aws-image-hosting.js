const request = require('request');
var winston = require('winston');
var aws = require('aws-sdk');
var sharp = require('sharp');
var async = require('async');
const _ = require('lodash');

const { VIEW_TYPE_OBJECT_DETAILS, VIEW_TYPE_GALLERY, VIEW_TYPE_TIMELINE } = require('../../config/views.config');

const BUCKET = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const S3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
    signatureVersion: 'v4',
    region: 'nyc3',
});

module.exports = {
    uploadToS3,
    signedUrlForPutObject,
    getAllIconsForTeam,
    hostImageLocatedAtRemoteURL,
    copyPageAssets,
    copyDatasetAsset,
    getAssetByKey,
};

function uploadToS3(key, response, callback) {

    var payload = {
        Bucket: BUCKET,
        Key: key,
        Body: response,
        ACL: 'public-read',
    };

    S3.upload(payload, function (err, data) {
        if (err) {
            winston.error('AWS S3 write stream error');
            return callback(err);
        }
        return callback(null, data);
    });
}

function _appendKeyToBucket(key) {
    return 'https://' + BUCKET + '.' + process.env.DO_S3_ENDPOINT + '/' + key;
}

function getImageFormat(headers) {
    const imageFormat = _.get(headers, 'content-type', '').split(';')[0].split('/')[1];

    return imageFormat === 'jpg' ? 'jpeg' : imageFormat;
}

function getImageAtUrl(remoteImageSourceURL, callback) {
    var options = {
        url: remoteImageSourceURL,
        encoding: null,
        timeout: 10000,
    };

    request.get(options, function (err, response, data) {
        if ((err && ['ENOTFOUND', 'ETIMEDOUT'].includes(err.code)) || !response) {
            winston.info(`returning url as null, since Could not read the remote image ${remoteImageSourceURL}: ${err}`);
            return callback(null, null);
        }

        if (err) {
            return callback(err, null);
        }

        const imageFormat = getImageFormat(response.headers);

        if (_.get(sharp, `format.${imageFormat}`)) {
            return callback(null, data);
        }

        return callback(null, null);
    });
}

function signedUrlForPutObject(key, fileType, callback) {
    if (fileType.indexOf('svg') >= 0) { //fix space when uploading svg error
        fileType = 'image/svg+xml';
    }

    var params = {
        Bucket: BUCKET,
        Key: key,
        ACL: 'public-read',
        ContentType: fileType,
    };

    S3.getSignedUrl('putObject', params, function (err, signedUrl) {
        callback(err, { putSignedUrl: signedUrl, publicUrl: _appendKeyToBucket(key) });
    });
}

function getAllIconsForTeam(teamSubdomain, callback) {
    S3.listObjects({
        Bucket: BUCKET,
        Prefix: teamSubdomain + '/assets/icon/',
        Marker: teamSubdomain + '/assets/icon/',
    }, function (err, data) {
        if (err) {
            callback(err);
        } else {
            var listOfUrls = [];
            var objects = data.Contents;

            for (var i = 0; i < objects.length; i++) {
                var objKey = objects[i].Key;

                var url = _appendKeyToBucket(objKey);

                listOfUrls.push(url);
            }
            callback(null, listOfUrls);
        }
    });

}

async function _proceedToStreamToHost(imageUrl, resize, folder, docPKey, callback) {
    try {
        const image = sharp(imageUrl);
        const metadata = await image.metadata();

        if (!metadata) {
            winston.info('returning url as null, since Could not read the remote image metadata');
            return callback();
        }

        const { width, hasAlpha } = metadata;
        const size = width < resize.size ? width : resize.size;
        const whiteColorRGB = { r: 255, g: 255, b: 255 };
        const background = hasAlpha ? { ...whiteColorRGB, alpha: 1 } : whiteColorRGB;
        const resizeConfig = { fit: 'contain', background };
        const resizedImage = resize.cropped ? image.resize(size, size, resizeConfig) : image.resize(size);
        const bufferedImage = await resizedImage.toBuffer();
        const keyName = `${folder}${resize.view}/${docPKey}`;

        uploadToS3(keyName, bufferedImage, callback);
    } catch (error) {
        winston.error('has err: ' + error);
        callback(error);
    }
}

function hostImageLocatedAtRemoteURL(folder, remoteImageSourceURL, overwrite, destinationFilenameSansExt, callback) {
    const recommendedSizes = [
        { size: 763, cropped: false, view: VIEW_TYPE_OBJECT_DETAILS },
        { size: 280, cropped: false, view: VIEW_TYPE_GALLERY },
        { size: 155, cropped: true, view: VIEW_TYPE_TIMELINE },
    ];

    getImageAtUrl(remoteImageSourceURL, function (err, imageBuffer) {
        if (err || !imageBuffer) {
            return callback(err);
        } else {
            async.each(recommendedSizes, function (size, cb) {
                if (overwrite) {
                    _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                } else {
                    const keyName = `${folder}${size.view}/${destinationFilenameSansExt}`;
                    const params = {
                        Bucket: BUCKET,
                        Key: keyName,
                    };
                    try {
                        S3.headObject(params, function (err, data) {
                            if (err || !data) {
                                winston.error('Head object err: ' + err);
                            } else {
                                winston.info(' File already uploaded and overwrite is false for object with pKey ' + destinationFilenameSansExt);
                                cb();
                            }
                        });
                    } catch (err) {

                        if (err.code === 'ENOTFOUND' || err.code === 'ETIMEDOUT') {
                            _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                        } else {
                            winston.error(err);
                            cb(err);
                        }
                    }
                }
            }, callback);
        }
    });

}

function copyPageAssets(args, callback) {
    var params = {
        Bucket: BUCKET,
        CopySource: BUCKET + '/' + args.subdomain + '/websites/' + args.originalWebsiteId + '/pages/' + args.originalPageId + '/' + args.imageType + '/' + args.fileName,
        Key: args.subdomain + '/websites/' + args.websiteId + '/pages/' + args.pageId + '/' + args.imageType + '/' + args.fileName,
        ACL: 'public-read',
    };

    S3.copyObject(params, function (err) {
        if (err) {
            winston.error('\' error copying page asset with params: \'', params);
            return callback(err);
        } else {
            callback();
        }
    });
}

async function copyDatasetAsset({ fromSubdomain, toSubdomain, fromId, toId, source, fileName }, callback) {
    const getUrl = ({ id, BUCKET, subdomain = fromSubdomain }) =>
        `${BUCKET ? `${BUCKET}/` : ''}${subdomain}/datasets/${id}/${source}/${fileName}`;
    const copySource = getUrl({ id: fromId, BUCKET });
    const key = getUrl({ id: toId, subdomain: toSubdomain });

    const params = {
        Bucket: BUCKET,
        CopySource: copySource,
        Key: key,
        ACL: 'public-read',
    };

    try {
        await S3.copyObject(params).promise();

        callback(null);
    } catch (e) {
        winston.error(`S3: error copying dataset asset: ${e}`);

        callback(e);
    }
}

function getAssetByKey(key, callback) {
    const params = {
        Bucket: BUCKET,
        Key: key,
    };

    S3.getObject(params, function (err, data) {
        if (err) {
            return callback(err);
        } else {
            return callback(null, data);
        }
    });
}
