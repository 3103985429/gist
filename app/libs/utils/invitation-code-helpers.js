const crypto = require('crypto');

/**
 * Returns a sha512 hash of the given string.
 * @param confirmation_code - Code to hash
 * @return
 */
module.exports.createHashWithString = confirmation_code => {
    // Create a new hash, and then update it and return the digest
    const hash = crypto.createHash('sha512');
    hash.update(confirmation_code);
    return hash.digest('hex');
};

/**
 * Generates a random short code.
 * @param codeSize - Size in bytes to generate a code
 * @return
 */
module.exports.generateRandomCode = codeSize => {
    // Get a random buffer
    const randomBuffer = crypto.randomBytes(codeSize);
    // Return the random buffer to string
    return randomBuffer.toString('hex');
};
