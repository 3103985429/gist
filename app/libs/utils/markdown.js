var fs = require('fs');
var marked = require('marked');

module.exports.getMarkdown = function (filePath) {
    var rawMarkdown = fs.readFileSync(filePath, 'utf-8');
    return rawMarkdown;
};

module.exports.parseMarkdown = function (markdown) {
    var options = {
        renderer: new marked.Renderer(),
        gfm: true,
        tables: true,
        breaks: false,
        pedantic: false,
        sanitize: true,
        smartLists: true,
        smartypants: false
    };
    marked.setOptions(options);

    var lexer = new marked.Lexer(options);
    var parsedBody = marked(markdown);
    return parsedBody;
};

module.exports.replaceCharts = function (parsed) {
    var chartRE = /((?:<p>)?&lt;chart data-source-key=&quot;(?=\S*)([a-zA-Z\d_-]*)&quot; data-chart-type=&quot;(?=\S*)([a-zA-Z\d-]*)&quot; class=&quot;chart&quot; data-revision=&quot;(\d+)&quot; \/&gt;(?:<\/p>)?)/g;
    var parsedReplaced = parsed.replace(chartRE, '<div id="$2" data-source-key="$2" data-chart-type="$3" class="$3 chart" data-revision="$4"></div>');
    return parsedReplaced;
};
