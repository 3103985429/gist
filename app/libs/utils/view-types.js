const CONTENT_ONLY_TYPE = 'contentOnly';

module.exports = {
    CONTENT_ONLY_TYPE,
};
