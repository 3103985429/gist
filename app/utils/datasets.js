const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
