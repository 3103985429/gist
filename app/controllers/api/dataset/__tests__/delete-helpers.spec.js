import { deleteSource } from '../delete-helpers';
import { hydrateDb, getId } from '../../../../../internals/testing/backend/utils';
import { pokemon1 } from '../../../../../internals/testing/backend/fixtures/datasets';


const pokemon1id = getId(pokemon1);

describe('Delete helpers', () => {
    beforeEach(async () => {
        await hydrateDb();
    });

    describe('deleteSource', () => {
        it('should delete source of dataset', async () => {
            const dataset = await deleteSource(pokemon1id);

            expect(dataset).toMatchSnapshot();
        });
    });
});
