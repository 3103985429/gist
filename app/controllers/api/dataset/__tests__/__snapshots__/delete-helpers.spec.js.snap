// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete helpers deleteSource should delete source of dataset 1`] = `Object {}`;
